src/storage/src/write_batch.rs

```diff
@@ -15,420 +15,159 @@
-        "Length of column {} not equals to other columns, expect {}, given {}",
-        name,
-        expect,
-        given
-    ))]
-    LenNotEquals {
-        name: String,
-        expect: usize,
-        given: usize,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display(
-        "Request is too large, max is {}, current is {}",
-        MAX_BATCH_SIZE,
-        num_rows
-    ))]
-    RequestTooLarge {
-        num_rows: usize,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Cannot align timestamp: {}", ts))]
-    TimestampOverflow { ts: i64 },
-
-    #[snafu(display("Failed to encode, source: {}", source))]
-    EncodeArrow {
-        backtrace: Backtrace,
-        source: ArrowError,
-    },
-
-    #[snafu(display("Failed to decode, source: {}", source))]
-    DecodeArrow {
-        backtrace: Backtrace,
-        source: ArrowError,
-    },
-
-    #[snafu(display("Failed to encode into protobuf, source: {}", source))]
-    EncodeProtobuf {
-        backtrace: Backtrace,
-        source: EncodeError,
-    },
-
-    #[snafu(display("Failed to decode from protobuf, source: {}", source))]
-    DecodeProtobuf {
-        backtrace: Backtrace,
-        source: DecodeError,
-    },
-
-    #[snafu(display("Failed to parse schema, source: {}", source))]
-    ParseSchema {
-        backtrace: Backtrace,
-        source: datatypes::error::Error,
-    },
-
-    #[snafu(display("Failed to decode, corrupted data {}", message))]
-    DataCorrupted {
-        message: String,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Failed to decode vector, source {}", source))]
-    DecodeVector {
-        backtrace: Backtrace,
-        source: datatypes::error::Error,
-    },
-
-    #[snafu(display("Failed to convert into protobuf struct, source {}", source))]
-    ToProtobuf {
-        source: proto::write_batch::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Failed to convert from protobuf struct, source {}", source))]
-    FromProtobuf {
-        source: proto::write_batch::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display(
-        "Failed to create default value for column {}, source: {}",
-        name,
-        source
-    ))]
-    CreateDefault {
-        name: String,
-        #[snafu(backtrace)]
-        source: datatypes::error::Error,
-    },
-}
+use store_api::storage::{OpType, WriteRequest};
 
-pub type Result<T> = std::result::Result<T, Error>;
+use crate::error::{
+    BatchMissingColumnSnafu, CreateDefaultSnafu, CreateRecordBatchSnafu, Error, HasNullSnafu,
+    LenNotEqualsSnafu, RequestTooLargeSnafu, Result, TypeMismatchSnafu, UnknownColumnSnafu,
+};
 
 /// Max number of updates of a write batch.
-const MAX_BATCH_SIZE: usize = 1_000_000;
+pub(crate) const MAX_BATCH_SIZE: usize = 1_000_000;
+
+/// Data of [WriteBatch].
+///
+/// We serialize this struct to the WAL instead of the whole `WriteBatch` to avoid
+/// storing unnecessary information.
```

Comment:
`WriteBatch` only has an extra row count. Is it very unnecessary?

Reply:
#777 adds another field to support deletion.

Reply:
#777 adds another field to support deletion.